File: src/Builders/StageBuilder.php
<?php

namespace ReactUbiquitous\Builders;

use ReactUbiquitous\DTOs\UIStageConfig;
use ReactUbiquitous\DTOs\UIPageConfig;
use ReactUbiquitous\Enums\UITheme;
use ReactUbiquitous\Enums\UIPageTransition;

final class StageBuilder
{
    private array $pages = [];
    private ?string $title = null;
    private ?string $description = null;
    private ?string $defaultPageId = null;
    private UITheme|string|null $theme = null;
    private ?string $className = null;
    private UIPageTransition|string|null $pageTransition = null;

    public function __construct(private readonly string $id) {}

    public static function make(string $id): self
    {
        return new self($id);
    }

    public function title(string $title): self
    {
        $clone = clone $this;
        $clone->title = $title;
        return $clone;
    }

    public function description(string $description): self
    {
        $clone = clone $this;
        $clone->description = $description;
        return $clone;
    }

    public function defaultPageId(string $pageId): self
    {
        $clone = clone $this;
        $clone->defaultPageId = $pageId;
        return $clone;
    }

    public function theme(UITheme|string $theme): self
    {
        $clone = clone $this;
        $clone->theme = $theme;
        return $clone;
    }

    public function className(string $className): self
    {
        $clone = clone $this;
        $clone->className = $className;
        return $clone;
    }

    public function pageTransition(UIPageTransition|string $transition): self
    {
        $clone = clone $this;
        $clone->pageTransition = $transition;
        return $clone;
    }

    public function addPage(UIPageConfig $page): self
    {
        $clone = clone $this;
        $clone->pages[] = $page;
        return $clone;
    }

    public function pages(array $pages): self
    {
        $clone = clone $this;
        $clone->pages = $pages;
        return $clone;
    }

    public function build(): UIStageConfig
    {
        return new UIStageConfig(
            id: $this->id,
            pages: $this->pages,
            title: $this->title,
            description: $this->description,
            defaultPageId: $this->defaultPageId,
            theme: $this->theme,
            className: $this->className,
            pageTransition: $this->pageTransition,
        );
    }
}

File: nuget/src/ReactUbiquitous.NuGet/Builders/StageBuilder.cs
using ReactUbiquitous.NuGet.Models;

namespace ReactUbiquitous.NuGet.Builders;

/// <summary>Provides a fluent builder for constructing <see cref="UIStageConfig"/> instances.</summary>
public class StageBuilder
{
    private readonly UIStageConfig _stage = new();

    /// <summary>Sets the unique stage identifier.</summary>
    public StageBuilder WithId(string id) { _stage.Id = id; return this; }

    /// <summary>Sets the stage title.</summary>
    public StageBuilder WithTitle(string title) { _stage.Title = title; return this; }

    /// <summary>Sets the stage description.</summary>
    public StageBuilder WithDescription(string description) { _stage.Description = description; return this; }

    /// <summary>Sets the identifier of the default page to display.</summary>
    public StageBuilder WithDefaultPageId(string pageId) { _stage.DefaultPageId = pageId; return this; }

    /// <summary>Sets the visual theme for the stage.</summary>
    public StageBuilder WithTheme(string theme) { _stage.Theme = theme; return this; }

    /// <summary>Sets additional CSS class names for the stage.</summary>
    public StageBuilder WithClassName(string className) { _stage.ClassName = className; return this; }

    /// <summary>Sets the page transition animation style.</summary>
    public StageBuilder WithPageTransition(string transition) { _stage.PageTransition = transition; return this; }

    /// <summary>Adds a page to the stage.</summary>
    public StageBuilder AddPage(UIPageConfig page) { (_stage.Pages ??= []).Add(page); return this; }

    /// <summary>Builds and returns the configured stage.</summary>
    public UIStageConfig Build() => _stage;
}
